python/cudf/cudf/_lib/pylibcudf/io/types.pyx

```diff
@@ -119,7 +158,102 @@ cdef class SourceInfo:
 
-        self.c_obj = move(source_info(c_host_buffers))
+        A homogeneous list of sinks (this can be a string filename,
+        bytes, or one of the Python I/O classes) to read from.
+
+        Mixing different types of sinks will raise a `ValueError`.
+    """
+
+    def __init__(self, list sinks):
+        cdef vector[data_sink *] data_sinks
+        cdef unique_ptr[data_sink] sink
+
+        cdef vector[string] paths
+
+        if not sinks:
+            raise ValueError("Need to pass at least one sink")
+
+        if isinstance(sinks[0], io.StringIO):
+            data_sinks.reserve(len(sinks))
+            for s in sinks:
+                if not isinstance(s, io.StringIO):
+                    raise ValueError("All sinks must be of the same type!")
+                self.sink_storage.push_back(
+                    unique_ptr[data_sink](new iobase_data_sink(s))
+                )
+                data_sinks.push_back(self.sink_storage.back().get())
+            self.c_obj = sink_info(data_sinks)
+        elif isinstance(sinks[0], io.IOBase):
+            data_sinks.reserve(len(sinks))
+            for s in sinks:
+                # Files opened in text mode expect writes to be str rather than
+                # bytes, which requires conversion from utf-8. If the underlying
+                # buffer is utf-8, we can bypass this conversion by writing
+                # directly to it.
+                if isinstance(s, io.TextIOBase):
+                    if codecs.lookup(s.encoding).name not in {
+                        "utf-8",
+                        "ascii",
+                    }:
+                        raise NotImplementedError(f"Unsupported encoding {s.encoding}")
+                    sink = move(unique_ptr[data_sink](new iobase_data_sink(s.buffer)))
+                elif isinstance(s, io.BytesIO):
+                    sink = move(unique_ptr[data_sink](new iobase_data_sink(s)))
+                else:
+                    raise ValueError("All sinks must be of the same type!")
+
+                self.sink_storage.push_back(
+                    move(sink)
+                )
+                data_sinks.push_back(self.sink_storage.back().get())
+            self.c_obj = sink_info(data_sinks)
+        elif isinstance(sinks[0], str):
+            paths.reserve(len(sinks))
+            for s in sinks:
+                if not isinstance(s, str):
+                    raise ValueError("All sinks must be of the same type!")
+                paths.push_back(<string> s.encode())
+            self.c_obj = sink_info(move(paths))
+        elif isinstance(sinks[0], os.PathLike):
+            paths.reserve(len(sinks))
+            for s in sinks:
+                if not isinstance(s, os.PathLike):
+                    raise ValueError("All sinks must be of the same type!")
+                paths.push_back(<string> os.path.expanduser(s).encode())
+            self.c_obj = sink_info(move(paths))
+        else:
```

Comment:
There is a significant amount of duplicated logic in this function that I think could be simplified if instead of having 4 branches of the if/else you had two branches that each branched twice. The first branch would handle the io types, while the second branch would handle str/PathLike. Also, you may be able to do a single pass check of `all(isinstance(s, type(sinks[0]) for s in sinks`, again with a bit of special handling for the io types case though since `type(sinks[0])` won't be general enough.

Just a suggestion though, simplify as much as you can but it's not blocking.

Reply:
Definitely better. I think it's worth one more small pass then we'll be good.